### ISSUE TITLE
Operator fails to register Informer watches due to "stream was reset: CANCEL" error

### ISSUE THREAD

**Issue by ivanchuchulski** [operator-framework/java-operator-sdk#2245]
Hello,

We're deploying an Operator, our Reconciler uses standalone dependent resources and implements the `EventSourceInitializer` interface.

We get the following error in the logs upon startup:
```(java)
2024-02-14 16:24:50.676 WARN  BaseConfigurationService.java (61) Configuration for reconciler 'myappsreconciler' was not found. Known reconcilers: None.
2024-02-14 16:24:50.696 INFO  BaseConfigurationService.java (83) Created configuration for reconciler com.mycompany.operator.MyAppReconciler with name myappsreconciler
2024-02-14 16:24:50.913 INFO  Operator.java                 (243) Registered reconciler: 'myappsreconciler' for resource: 'class com.mycompany.operator.MyApp' for namespace(s): [all namespaces]
2024-02-14 16:24:50.914 INFO  Operator.java                 (147) Operator SDK 4.4.4 (commit: 609a55a) built on 2023-09-19T10:58:23.000+0000 starting...
2024-02-14 16:24:50.915 INFO  Operator.java                 (153) Client version: 6.7.2
2024-02-14 16:24:50.917 INFO  Controller.java               (334) Starting 'myappsreconciler' controller for reconciler: com.mycompany.operator.MyAppReconciler, resource: com.mycompany.operator.MyApp
2024-02-14 16:24:51.396 WARN  VersionUsageUtils.java        (60) The client is using resource type 'certificates' with unstable version 'v1alpha1'
2024-02-14 16:24:51.403 WARN  VersionUsageUtils.java        (60) The client is using resource type 'virtualservices' with unstable version 'v1beta1'
2024-02-14 16:24:51.401 WARN  VersionUsageUtils.java        (60) The client is using resource type 'destinationrules' with unstable version 'v1beta1'
2024-02-14 16:24:51.478 WARN  VersionUsageUtils.java        (60) The client is using resource type 'peerauthentications' with unstable version 'v1beta1'
2024-02-14 16:24:51.502 WARN  VersionUsageUtils.java        (60) The client is using resource type 'gateways' with unstable version 'v1beta1'
2024-02-14 16:25:01.903 ERROR Reflector.java                (153) listSyncAndWatch failed for v1/secrets, will stop
java.util.concurrent.CompletionException: okhttp3.internal.http2.StreamResetException: stream was reset: CANCEL
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:636) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162) ~[?:?]
        at io.fabric8.kubernetes.client.http.ByteArrayBodyHandler.onBodyDone(ByteArrayBodyHandler.java:46) ~[kubernetes-client-api-6.7.2.jar:?]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162) ~[?:?]
        at io.fabric8.kubernetes.client.okhttp.OkHttpClientImpl$OkHttpAsyncBody.doConsume(OkHttpClientImpl.java:142) ~[kubernetes-httpclient-okhttp-6.7.2.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
        at java.lang.Thread.run(Thread.java:833) [?:?]
Caused by: okhttp3.internal.http2.StreamResetException: stream was reset: CANCEL
        at okhttp3.internal.http2.Http2Stream$FramingSource.read(Http2Stream.java:420) ~[okhttp-3.12.12.jar:?]
        at okhttp3.internal.http2.Http2Codec$StreamFinishingSource.read(Http2Codec.java:205) ~[okhttp-3.12.12.jar:?]
        at okio.RealBufferedSource.read(RealBufferedSource.java:47) ~[okio-1.15.0.jar:?]
        at okio.RealBufferedSource.exhausted(RealBufferedSource.java:57) ~[okio-1.15.0.jar:?]
        at okio.InflaterSource.refill(InflaterSource.java:102) ~[okio-1.15.0.jar:?]
        at okio.InflaterSource.read(InflaterSource.java:62) ~[okio-1.15.0.jar:?]
        at okio.GzipSource.read(GzipSource.java:80) ~[okio-1.15.0.jar:?]
        at okio.RealBufferedSource.exhausted(RealBufferedSource.java:57) ~[okio-1.15.0.jar:?]
        at io.fabric8.kubernetes.client.okhttp.OkHttpClientImpl$OkHttpAsyncBody.doConsume(OkHttpClientImpl.java:131) ~[kubernetes-httpclient-okhttp-6.7.2.jar:?]
        ... 3 more
2024-02-14 16:25:02.132 ERROR InformerWrapper.java          (93) Informer startup error. Operator will be stopped. Informer: v1/secrets
java.util.concurrent.ExecutionException: okhttp3.internal.http2.StreamResetException: stream was reset: CANCEL
        at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396) ~[?:?]
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096) ~[?:?]
        at io.javaoperatorsdk.operator.processing.event.source.informer.InformerWrapper.start(InformerWrapper.java:87) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.processing.event.source.informer.InformerManager.lambda$start$0(InformerManager.java:62) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.lambda$executeAndWaitForAllToComplete$0(ExecutorServiceManager.java:70) ~[operator-framework-core-4.4.4.jar:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
        at java.lang.Thread.run(Thread.java:833) [?:?]
Caused by: okhttp3.internal.http2.StreamResetException: stream was reset: CANCEL
        at okhttp3.internal.http2.Http2Stream$FramingSource.read(Http2Stream.java:420) ~[okhttp-3.12.12.jar:?]
        at okhttp3.internal.http2.Http2Codec$StreamFinishingSource.read(Http2Codec.java:205) ~[okhttp-3.12.12.jar:?]
        at okio.RealBufferedSource.read(RealBufferedSource.java:47) ~[okio-1.15.0.jar:?]
        at okio.RealBufferedSource.exhausted(RealBufferedSource.java:57) ~[okio-1.15.0.jar:?]
        at okio.InflaterSource.refill(InflaterSource.java:102) ~[okio-1.15.0.jar:?]
        at okio.InflaterSource.read(InflaterSource.java:62) ~[okio-1.15.0.jar:?]
        at okio.GzipSource.read(GzipSource.java:80) ~[okio-1.15.0.jar:?]
        at okio.RealBufferedSource.exhausted(RealBufferedSource.java:57) ~[okio-1.15.0.jar:?]
        at io.fabric8.kubernetes.client.okhttp.OkHttpClientImpl$OkHttpAsyncBody.doConsume(OkHttpClientImpl.java:131) ~[kubernetes-httpclient-okhttp-6.7.2.jar:?]
        ... 3 more
...(multiple retries)
2024-02-14 16:25:02.895 ERROR MyAppOperator.java(42) Failed to start Operator
io.javaoperatorsdk.operator.OperatorException: Error starting operator
        at io.javaoperatorsdk.operator.Operator.start(Operator.java:166) ~[operator-framework-core-4.4.4.jar:?]
        at myapp-operator.operator.MyAppOperator.main(MyAppOperator.java:35) [my-app-operator-1.0.0-SNAPSHOT.jar:?]
Caused by: io.javaoperatorsdk.operator.OperatorException: io.javaoperatorsdk.operator.OperatorException: io.javaoperatorsdk.operator.OperatorException: Couldn't start source io.javaoperatorsdk.operator.processing.event.source.informer.InformerEventSource#2002720253
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.lambda$executeAndWaitForAllToComplete$2(ExecutorServiceManager.java:81) ~[operator-framework-core-4.4.4.jar:?]
        at java.util.ArrayList.forEach(ArrayList.java:1511) ~[?:?]
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.executeAndWaitForAllToComplete(ExecutorServiceManager.java:76) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.boundedExecuteAndWaitForAllToComplete(ExecutorServiceManager.java:56) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.ControllerManager.start(ControllerManager.java:42) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.Operator.start(Operator.java:161) ~[operator-framework-core-4.4.4.jar:?]
        ... 1 more
Caused by: io.javaoperatorsdk.operator.OperatorException: io.javaoperatorsdk.operator.OperatorException: Couldn't start source io.javaoperatorsdk.operator.processing.event.source.informer.InformerEventSource#2002720253
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.lambda$executeAndWaitForAllToComplete$2(ExecutorServiceManager.java:81) ~[operator-framework-core-4.4.4.jar:?]
        at java.util.ArrayList.forEach(ArrayList.java:1511) ~[?:?]
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.executeAndWaitForAllToComplete(ExecutorServiceManager.java:76) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.boundedExecuteAndWaitForAllToComplete(ExecutorServiceManager.java:56) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.processing.event.EventSourceManager.start(EventSourceManager.java:79) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.processing.Controller.start(Controller.java:342) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.ControllerManager.lambda$start$0(ControllerManager.java:43) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.lambda$executeAndWaitForAllToComplete$0(ExecutorServiceManager.java:70) ~[operator-framework-core-4.4.4.jar:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: io.javaoperatorsdk.operator.OperatorException: Couldn't start source io.javaoperatorsdk.operator.processing.event.source.informer.InformerEventSource#2002720253
        at io.javaoperatorsdk.operator.processing.event.EventSourceManager.startEventSource(EventSourceManager.java:130) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.lambda$executeAndWaitForAllToComplete$0(ExecutorServiceManager.java:70) ~[operator-framework-core-4.4.4.jar:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: io.javaoperatorsdk.operator.OperatorException: io.javaoperatorsdk.operator.OperatorException: Couldn't start informer for secrets/v1 resources
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.lambda$executeAndWaitForAllToComplete$2(ExecutorServiceManager.java:81) ~[operator-framework-core-4.4.4.jar:?]
        at java.util.ArrayList.forEach(ArrayList.java:1511) ~[?:?]
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.executeAndWaitForAllToComplete(ExecutorServiceManager.java:76) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.boundedExecuteAndWaitForAllToComplete(ExecutorServiceManager.java:56) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.processing.event.source.informer.InformerManager.start(InformerManager.java:59) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.processing.event.source.informer.ManagedInformerEventSource.start(ManagedInformerEventSource.java:80) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.processing.event.NamedEventSource.start(NamedEventSource.java:27) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.processing.event.EventSourceManager.startEventSource(EventSourceManager.java:125) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.lambda$executeAndWaitForAllToComplete$0(ExecutorServiceManager.java:70) ~[operator-framework-core-4.4.4.jar:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: io.javaoperatorsdk.operator.OperatorException: Couldn't start informer for secrets/v1 resources
        at io.javaoperatorsdk.operator.processing.event.source.informer.InformerWrapper.start(InformerWrapper.java:110) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.processing.event.source.informer.InformerManager.lambda$start$0(InformerManager.java:62) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.lambda$executeAndWaitForAllToComplete$0(ExecutorServiceManager.java:70) ~[operator-framework-core-4.4.4.jar:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: io.javaoperatorsdk.operator.OperatorException: java.util.concurrent.ExecutionException: okhttp3.internal.http2.StreamResetException: stream was reset: CANCEL
        at io.javaoperatorsdk.operator.processing.event.source.informer.InformerWrapper.start(InformerWrapper.java:94) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.processing.event.source.informer.InformerManager.lambda$start$0(InformerManager.java:62) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.lambda$executeAndWaitForAllToComplete$0(ExecutorServiceManager.java:70) ~[operator-framework-core-4.4.4.jar:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: java.util.concurrent.ExecutionException: okhttp3.internal.http2.StreamResetException: stream was reset: CANCEL
        at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396) ~[?:?]
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096) ~[?:?]
        at io.javaoperatorsdk.operator.processing.event.source.informer.InformerWrapper.start(InformerWrapper.java:87) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.processing.event.source.informer.InformerManager.lambda$start$0(InformerManager.java:62) ~[operator-framework-core-4.4.4.jar:?]
        at io.javaoperatorsdk.operator.api.config.ExecutorServiceManager.lambda$executeAndWaitForAllToComplete$0(ExecutorServiceManager.java:70) ~[operator-framework-core-4.4.4.jar:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: okhttp3.internal.http2.StreamResetException: stream was reset: CANCEL
        at okhttp3.internal.http2.Http2Stream$FramingSource.read(Http2Stream.java:420) ~[okhttp-3.12.12.jar:?]
        at okhttp3.internal.http2.Http2Codec$StreamFinishingSource.read(Http2Codec.java:205) ~[okhttp-3.12.12.jar:?]
        at okio.RealBufferedSource.read(RealBufferedSource.java:47) ~[okio-1.15.0.jar:?]
        at okio.RealBufferedSource.exhausted(RealBufferedSource.java:57) ~[okio-1.15.0.jar:?]
        at okio.InflaterSource.refill(InflaterSource.java:102) ~[okio-1.15.0.jar:?]
        at okio.InflaterSource.read(InflaterSource.java:62) ~[okio-1.15.0.jar:?]
        at okio.GzipSource.read(GzipSource.java:80) ~[okio-1.15.0.jar:?]
        at okio.RealBufferedSource.exhausted(RealBufferedSource.java:57) ~[okio-1.15.0.jar:?]
        at io.fabric8.kubernetes.client.okhttp.OkHttpClientImpl$OkHttpAsyncBody.doConsume(OkHttpClientImpl.java:131) ~[kubernetes-httpclient-okhttp-6.7.2.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]

```
I can see that the error is happening during the registration of the watches in the Informers, but it's from OkHttp error, which I believe is the underlying HTTP server used in Fabric8 client.

Have you ever encountered these exception, as they should be crusial ones:
```(java)
2024-02-19 18:31:34.425 ERROR Reflector.java                (153) listSyncAndWatch failed for v1/secrets, will stop
java.util.concurrent.CompletionException: okhttp3.internal.http2.StreamResetException: stream was reset: CANCEL
```
and what is the meaning of the stream reset? The types of dependent resources we use are a lot of the standard K8s resources, like `ServiceAccount`, `Secret`, `ConfigMap`, `Deployment`, as well as some Istio resources. However, the error is always thrown for the Secrets watch. It's obviously not a RBAC one, because we've fixed those :) . Could it be related to some network policies in the cluster?

#### Environment

Kyma provisioned Gardener Kubernetes cluster with Istio installed; I believe cloud provider is Azure
Java Operator SDK version 4.4.4
Java version 17
Kubernetes version 1.27.8
Istio version 1.20.3

**Comment by csviri** [comment]
Hi @ivanchuchulski , this seems to be an issue with the client. What I would propose is to upgrade to a newer version of the SDK,  it contains also a newer version of the fabric8 client. 
What you can do is to replace the underlying http client, usually we recommend using Vert.x client.
See in the integration tests, we are testing with other clients too: https://github.com/operator-framework/java-operator-sdk/blob/69711cf346d75331a61616bcfe2fd7015c93434b/.github/workflows/integration-tests.yml#L12-L12

The retry might be a matter of a configuration, worth to experiment with:
https://github.com/operator-framework/java-operator-sdk/blob/69711cf346d75331a61616bcfe2fd7015c93434b/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/config/ConfigurationService.java#L230-L230

cc @manusa @shawkins have you seen this before? thank you!

**Comment by manusa** [comment]
Haven't seen this (specifically) before.
The timeline is kind of suspicious, since it happens exactly 10 seconds after the start-up (looks like it might be some sort of timeout). Everything points in the direction that the client can't establish a connection (or a websocket connection) with the cluster.
Have you tried this with a different cluster?

**Comment by ivanchuchulski** [comment]
We went over out configurations and it turned out that our Operator didn't have proper filtering in the EventSources, which resulted in a lot of processing on the initial loading during the registering of the underlying watches. Ultimately some of them probably failed with some network timeout.

We saw this type of behavior for now only on one cluster which had a lot more resources - around 1200+ in total of Service, ConfigMap and so on resources.

We mitigated the probable network error for now by adding label selectors to our dependent resources configurations. For example with this change it still takes about 4-5 seconds for the Reconciler to process the watches:
```(java)
2024-02-13 13:40:33.978 INFO  Controller.java               (334) Starting 'myappreconciler' controller for reconciler: ...
...
2024-02-13 13:40:37.498 INFO  Controller.java               (346) 'myappreconciler' controller started
```

**Comment by csviri** [comment]
So if I understood, this is kinda timouted? The 1200+ resources does not seems to be lot. There are operators written in JOSDK which handle 400k+ resources, and we don't know about any problems.

**Comment by ivanchuchulski** [comment]
Hi, 
That's the timeout due to Pod resources exhaustion is my best guess at the moment. It was checked by a cluster admin that the request doesn't get cancelled by K8s API side. What we observed was a high CPU usage on the Operator Pod, but from increasing the log level, I'm not able to pinpoint anything particular, since also I'm not aware with the general flow of the watch registration: do they go parallel at some point and because let's say for the first one immediately events begin to be received and that overwhelms the processing or is it something else.

So we went with the approach I described in my last reply and for now it seems mitigated. Overall that is a change, that we were planing to do anyways, but it was kind of forced by this problem.

One more thing worth mentioning: you say 400k+ resources handled is that by one Operator and one reconciler and what are the CPU usages at that point for the Pod? I guess we have that point to investigate also: should we set CPU limits to the Operator container or not. I saw polarizing opinions about that. Since we don't have any data to back it up, like Prometheus or some other scraper, I can't really determine what we should set.
